### ISSUE TITLE
Wrong 32bit library path in RPM based distributions

### ISSUE THREAD

**Issue by VortexAcherontic** [flightlessmango/MangoHud#424]
Hey there initially I thought this is a pressure vessel issue like described in the pinned issue #369 but I found out that with the current SoldierRuntime (0.20201124.0) and Proton 5.13-3 MangoHud works fine on games using Proton + DXVK if they are 64bit games.

Also I assume this is not a pressure vessel or Steam Runtime issue since I can reproduce the same behavior by using native games or games running via Lutris + Wine

Unfortunately I have no native 32bit Vulkan game available so this is the only case which remains untested

I also checked the MangoHud installation and the 32bit sos seems to be there I prepared some logs which hopefully help you narrowing down this issue. Or I am just missing something on my end.

- ❌ [Gothic 3 - Wine - DXVK - 32bit - Log](https://pastebin.com/zuy0HBj4)
- ✔ [The Witcher 3 - Wine - DXVK - 64bit - Log](https://pastebin.com/Mt33HrYs)
- ❌ [Darksburg - Proton - OpenGL - 64bit - Proton Log - Tried --dlsym as well to no avail here](https://pastebin.com/KtzcVygK)
- ✔ [Darksburg - Proton - DXVK - 64bit - Proton Log](https://pastebin.com/XkDc8TEU)
- ❌ [Kao: The Kangaroo: Round 2 - Proton - OpenGL - 32bit - Log - Tried --dlsym here too to no avail](https://pastebin.com/Mc5LUUeD) Also this one [worked](https://youtu.be/ibRWoUbg_ao?t=578) with an older version of Proton and MangoHud 
- ❌ [Fitzquake - Native - OpenGL - 32bit - Log](https://pastebin.com/c9wYYmL1)
- ✔ [GLXgears - Native - OpenGL - 64bit - Log](https://pastebin.com/m2WPWKhX)
- ✔ [VKCube - Native - Vulkan - 64bit - Log](https://pastebin.com/JWxaSHJE)

Specs:
MangoHud: 0.6.1
OS: openSUSE Tumbleweed
Kernel: 5.9.12
GPU: GTX 1080
Driver: 455.46.1

**Comment by Leopard1907** [comment]
That is still a runtime issue. 32 bit layers doesn't work with runtime.

https://gitlab.steamos.cloud/steamrt/steam-runtime-tools/-/issues/39

**Comment by flightlessmango** [comment]
I would say this still falls under #369

**Comment by VortexAcherontic** [comment]
@flightlessmango This would not explain the native games who did not worked as well as any game running via Lutris withput Pressure Vessel. But it may very well the case that I just missed something on my end so I thinkt for now it is the better dessicion to close this issue until I can verify it is indeed an issue and not me just forget about something 😄 

@Leopard1907 Which version of wine did you use? I had lutris-5.21 running in my tests in which the Hud did not worked.

**Comment by Leopard1907** [comment]
Lutris 5.21 is the build i used.

**Comment by flightlessmango** [comment]
Try using something like `glxspheres32`, comes with the 32bit virtualgl package on arch

**Comment by flightlessmango** [comment]
Or try this [vkcube32](https://cdn.discordapp.com/attachments/641713920247332877/752546590979260435/vkcube32)

**Comment by VortexAcherontic** [comment]
I guess something is off with my installation but I can not narrow down what it is:

````
mangohud ./vkcube32 
ERROR: ld.so: object '/usr/lib/mangohud/$LIB/libMangoHud.so' from LD_PRELOAD cannot be preloaded (wrong ELF class: ELFCLASS64): ignored.
failed to initialize wayland, falling back to xcb
2 physical devices
vendor id 10de, device name GeForce GTX 1080
````

**Comment by flightlessmango** [comment]
Are you using the release package?

**Comment by VortexAcherontic** [comment]
Yes I do

**Comment by DadSchoorse** [comment]
The issue probably is that the mangohud install script uses `/lib` for 64bit, but on opensuse/fedora it's for 32bit (and `/lib64` for 64bit).

**Comment by VortexAcherontic** [comment]
It was working fine with any prior release until recently but also Manghud has symlinks for this to address this:

````
/usr/lib/mangohud> LC_ALL=C ll
total 24
lrwxrwxrwx 1 root root   5 Dec 10 00:30 i386-linux-gnu -> lib32
lrwxrwxrwx 1 root root   5 Dec 10 00:30 i686 -> lib32
lrwxrwxrwx 1 root root   5 Dec 10 00:30 i686-linux-gnu -> lib32
drwxr-xr-x 1 root root 184 Dec 10 00:30 lib
drwxr-xr-x 1 root root  68 Dec 10 00:30 lib32
lrwxrwxrwx 1 root root   3 Dec 10 00:30 lib64 -> lib
drwxr-xr-x 1 root root  20 Dec 10 00:30 tls
lrwxrwxrwx 1 root root   3 Dec 10 00:30 x86_64 -> lib
lrwxrwxrwx 1 root root   3 Dec 10 00:30 x86_64-linux-gnu -> lib
````

**Comment by DadSchoorse** [comment]
Yes, and those symlinks are wrong on rpm based distros. `lib` should be a symlink to `lib32`, not `lib64`.

**Comment by VortexAcherontic** [comment]
You're right sir! That did the trick! Thank you.

I moved the folder in /usr/lib/mangohud to be as follows

lib for 32bit
lib64 for 64bit

and now things work as expected.

![Screenshot_20201211_110914](https://user-images.githubusercontent.com/11244092/101890682-4fd40300-3ba1-11eb-8641-ea5ec0430484.png)

@flightlessmango fyi.

**Comment by AngryPenguinPL** [comment]
Yes, same issue on OpenMandriva. 
Installed both x86_64 and i686 packages.
x86_64 provides /usr/lib64/mangohud/ and contains both libMangoHud.so and libMangoHud.dlsym.so
while i686 provides /usr/lib/mangohud/ and contains both libMangoHud.so and libMangoHud.dlsym.so
 but overlay on 32-bit app still won\t work

**Comment by dagolinuxoid** [comment]
Rather than creating a new issue I decided to put my question here:
I'm trying to get mangohud overlay to work on Guilty Gear AC + R game which I'm launching via wine but it doesn't work (I do have working overlay when I play dota 2 through steam though) . Elementary OS Hera, there is the result:
[![mangohud.png](https://i.postimg.cc/RVTjzgV3/mangohud.png)](https://postimg.cc/TppkVqbG)
I tried to create those wine-explorer and other not founded files with a content of Mangohud.cfg but it didn't help.

**Comment by VortexAcherontic** [comment]
The configs are irrelevant and just extra places MangoHud looks for config files in case you want different HUD elements for different applications here and also this seems not to belong to this issue.
You can however instead try:

`mangohud --dlsym wine GGXXACPR_Win.exe`

**Comment by dagolinuxoid** [comment]
`--dlsym` flag did the trick thx vortex

**Comment by Hasshu** [comment]
Any updates on this? Fedora and openSUSE aren't exactly some random hobby distros.

**Comment by VortexAcherontic** [comment]
In the openSUSE repos this issue is currently worked around by some build time modifications: https://build.opensuse.org/package/show/games:tools/mangohud
Maybe you can talk to the fedora maintainers and see if they could make use of it as well so some does not need to wait for the devs to catch up.

**Comment by Hasshu** [comment]
@VortexAcherontic According to [this](https://bodhi.fedoraproject.org/updates/FEDORA-2021-ee587d803a), the 32-bit package should install "automagically" on Fedora these days... yet I just found out that it didn't happen. Maybe that doesn't work when MangoHud is among the first packages to be installed on a fresh system. Or maybe the maintainer is a Muggle.

Either way, it's always a good idea to fix such issues upstream.

**Comment by Fatmice** [comment]
For fedora, I needed to fix all of the symlinks and directory names into these
```
tree -s /usr/lib/mangohud
/usr/lib/mangohud
├── [          3]  i386-linux-gnu -> lib
├── [          3]  i686 -> lib
├── [          3]  i686-linux-gnu -> lib
├── [       4096]  lib
│   ├── [      21044]  libMangoHud_dlsym.so
│   └── [    2104808]  libMangoHud.so
├── [       4096]  lib64
│   ├── [          6]  i386-linux-gnu -> ../lib
│   ├── [          1]  i686-linux-gnu -> .
│   ├── [          6]  lib32 -> ../lib
│   ├── [      22464]  libMangoHud_dlsym.so
│   ├── [    1983192]  libMangoHud.so
│   ├── [          6]  tls -> ../tls
│   ├── [          1]  x86_64 -> .
│   └── [          1]  x86_64-linux-gnu -> .
├── [       4096]  tls
│   ├── [          6]  i686 -> ../lib
│   └── [          9]  x86_64 -> ../lib64/
├── [          5]  x86_64 -> lib64
└── [          5]  x86_64-linux-gnu -> lib64
```

For comparison, these are made by the installation script
```
tree -s /usr/lib/mangohud/
/usr/lib/mangohud/
├── [          5]  i386-linux-gnu -> lib32
├── [          5]  i686 -> lib32
├── [          5]  i686-linux-gnu -> lib32
├── [       4096]  lib
│   ├── [          8]  i386-linux-gnu -> ../lib32
│   ├── [          1]  i686-linux-gnu -> .
│   ├── [          8]  lib32 -> ../lib32
│   ├── [      22464]  libMangoHud_dlsym.so
│   ├── [    1983192]  libMangoHud.so
│   ├── [          6]  tls -> ../tls
│   ├── [          1]  x86_64 -> .
│   └── [          1]  x86_64-linux-gnu -> .
├── [       4096]  lib32
│   ├── [      21044]  libMangoHud_dlsym.so
│   └── [    2104808]  libMangoHud.so
├── [          3]  lib64 -> lib
├── [       4096]  tls
│   ├── [          8]  i686 -> ../lib32
│   └── [          6]  x86_64 -> ../lib
├── [          3]  x86_64 -> lib
└── [          3]  x86_64-linux-gnu -> lib
```

**Comment by AlexFolland** [comment]
FYI, from my testing of this today and yesterday, mangohud works perfectly fine with 32-bit games in wine on Manjaro unstable using the `lib32-mangohud-git` AUR package.  I used `mangohud --dlsym wine '/home/alex/.wine/drive_c/games/fix it felix
 jr/FIFJ.exe'` and that showed the mangohud overlay correctly here.

Therefore, I believe this issue ticket as it is currently titled ("MangoHud does not work in 32bit games Proton/Wine/Native DXVK/OpenGL" as of writing this comment) is not necessarily entirely valid, as there are situations where it works as expected with 32-bit software running in wine.

Is the issue resolved in Fedora and OpenSUSE yet?  If not, does that really reflect poorly on the mangohud project?  If not, can this ticket be closed?  I ask because this ticket's contents misled me when I was trying to get mangohud to work on Manjaro yesterday and all I needed to do was install the `lib32-mangohud-git` AUR package and not do some surgical directory tree and symlink manipulation.
